core/src/main/java/io/lionweb/lioncore/java/serialization/AbstractSerialization.java

```diff
@@ -0,0 +1,561 @@
+
+  /**
+   * This guides what we do when deserializing a sub-tree and not being able to resolve the parent.
+   */
+  protected UnavailableNodePolicy unavailableParentPolicy = UnavailableNodePolicy.THROW_ERROR;
+
+  /**
+   * This guides what we do when deserializing a sub-tree and not being able to resolve the
+   * children.
+   */
+  protected UnavailableNodePolicy unavailableChildrenPolicy = UnavailableNodePolicy.THROW_ERROR;
+
+  /**
+   * This guides what we do when deserializing a sub-tree and not being able to resolve a reference
+   * target.
+   */
+  protected UnavailableNodePolicy unavailableReferenceTargetPolicy =
+      UnavailableNodePolicy.THROW_ERROR;
+
+  protected AbstractSerialization() {
+    // prevent public access
+    classifierResolver = new ClassifierResolver();
+    instantiator = new Instantiator();
+    primitiveValuesSerialization = new PrimitiveValuesSerialization();
+    instanceResolver = new LocalClassifierInstanceResolver();
+  }
+
+  //
+  // Configuration
+  //
+
+  public ClassifierResolver getClassifierResolver() {
+    return classifierResolver;
+  }
+
+  public Instantiator getInstantiator() {
+    return instantiator;
+  }
+
+  public PrimitiveValuesSerialization getPrimitiveValuesSerialization() {
+    return primitiveValuesSerialization;
+  }
+
+  public LocalClassifierInstanceResolver getInstanceResolver() {
+    return instanceResolver;
+  }
+
+  public void enableDynamicNodes() {
+    instantiator.enableDynamicNodes();
+    primitiveValuesSerialization.enableDynamicNodes();
+  }
+
+  public @Nonnull UnavailableNodePolicy getUnavailableParentPolicy() {
+    return this.unavailableParentPolicy;
+  }
+
+  public @Nonnull UnavailableNodePolicy getUnavailableReferenceTargetPolicy() {
+    return this.unavailableReferenceTargetPolicy;
+  }
+
+  public @Nonnull UnavailableNodePolicy getUnavailableChildrenPolicy() {
+    return this.unavailableChildrenPolicy;
+  }
+
+  public void setUnavailableParentPolicy(@Nonnull UnavailableNodePolicy unavailableParentPolicy) {
+    Objects.requireNonNull(unavailableParentPolicy);
+    this.unavailableParentPolicy = unavailableParentPolicy;
+  }
+
+  public void setUnavailableChildrenPolicy(
+      @Nonnull UnavailableNodePolicy unavailableChildrenPolicy) {
+    Objects.requireNonNull(unavailableChildrenPolicy);
+    this.unavailableChildrenPolicy = unavailableChildrenPolicy;
+  }
+
+  public void setUnavailableReferenceTargetPolicy(
+      @Nonnull UnavailableNodePolicy unavailableReferenceTargetPolicy) {
+    Objects.requireNonNull(unavailableReferenceTargetPolicy);
+    this.unavailableReferenceTargetPolicy = unavailableReferenceTargetPolicy;
+  }
+
+  public void registerLanguage(Language language) {
+    getClassifierResolver().registerLanguage(language);
+    getPrimitiveValuesSerialization().registerLanguage(language);
+  }
+
+  //
+  // Serialization to chunk
+  //
+
+  public SerializedChunk serializeTreeToSerializationBlock(ClassifierInstance<?> root) {
```

Comment:
Why do we have three times almost the same method? Seems like big API surface.

Reply:
I agree they are useful. The question is whether we have to put them into this API, or externalize them into helper methods -- (mostly) a matter of taste.

If they are API, we have to test all three of them for each implementation -- externalized ones only one time.